kimchi/src/circuits/gate.rs

```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
 /// analogously using `joint_combiner`.
 ///
 /// This function computes that combined value.
-pub fn combine_table_entry<'a, F: Field, I: DoubleEndedIterator<Item = &'a F>>(
+pub fn combine_table_entry<
+    'a,
+    F: 'a + Zero + One + Add + Mul + Clone,
+    I: DoubleEndedIterator<Item = &'a F>,
+>(
     joint_combiner: F,
     v: I,
 ) -> F {
-    v.rev().fold(F::zero(), |acc, x| joint_combiner * acc + x)
+    v.rev()
+        .fold(F::zero(), |acc, x| joint_combiner.clone() * acc + x.clone())
 }
 
-impl<F: Field> SingleLookup<F> {
+impl<F: Copy> SingleLookup<F> {
     /// Evaluate the linear combination specifying the lookup value to a field element.
-    pub fn evaluate<G: Fn(LocalPosition) -> F>(&self, eval: G) -> F {
+    pub fn evaluate<K: Zero + Add + Mul<F, Output = K>, G: Fn(LocalPosition) -> K>(
```

Comment:
in general it's nicer to move these bounds to a `where` clause at the end of the function signature